Framework/Core/src/LifetimeHelpers.cxx

```diff
@@ -26,27 +29,68 @@ namespace o2
 namespace framework
 {
 
-ExpirationHandler::Checker LifetimeHelpers::expireNever()
+namespace
 {
-  return [](int64_t) -> bool { return false; };
+size_t getCurrentTime()
+{
+  auto now = std::chrono::system_clock::now();
+  auto duration = now.time_since_epoch();
+  return std::chrono::duration_cast<std::chrono::microseconds>(duration).count();
+}
 }
 
-ExpirationHandler::Checker LifetimeHelpers::expireAlways()
+ExpirationHandler::Creator LifetimeHelpers::dataDrivenCreation()
 {
-  return [](int64_t) -> bool { return true; };
+  return [](TimesliceIndex&) -> void {
+    return;
+  };
 }
 
-namespace
+ExpirationHandler::Creator LifetimeHelpers::timeDrivenCreation(std::chrono::microseconds period)
 {
-auto getCurrentTime()
+  auto start = getCurrentTime();
+  auto last = std::make_shared<decltype(start)>(start);
+  // FIXME: should create timeslices when period expires....
+  return [last, period](TimesliceIndex& index) -> void {
+    // Nothing to do if the time has not expired yet.
+    auto current = getCurrentTime();
+    auto delta = current - *last;
+    if (delta < period.count()) {
+      return;
+    }
+    // We first check if the current time is not already present
+    // FIXME: this should really be done by query matching? Ok
+    //        for now to avoid duplicate entries.
+    for (size_t i = 0; i < index.size(); ++i) {
+      TimesliceSlot slot{ i };
+      if (index.isValid(slot) == false) {
+        continue;
+      }
+      if (index.getTimesliceForSlot(slot).value == current) {
+        return;
+      }
+    }
+    // If we are here the timer has expired and a new slice needs
+    // to be created.
+    *last = current;
+    data_matcher::VariableContext newContext;
+    auto slot = index.replaceLRUWith(newContext);
+    index.associate(TimesliceId{ current }, slot);
```

Comment:
I would actually use a counter for creating the time slice id. That would implement the same functionality as the `DataSourceDevice`. Though for sure not the final solution, it would be better than just a number derived from the time which will probably make the matching difficult (if the matching in the input cache lines is based on the ids). In the end we would need a central instance which creates time slice ids for a given time/period. For the moment I would simply stick to the counter as a simple mockup.

Reply:
agreed. I am refactoring this to add both the counter (which will be the default) or a timer.

Reply:
yes, I think the `Lifetime::Enumerate` option is a good solution